### ISSUE TITLE
Express your opinion: shift in the package mindset

### ISSUE THREAD

**Issue by lulupointu** [lulupointu/vrouter#13]
Hi,

I am considering changing this package quite a bit. I know that people are already using it and rely on it but it has always been obvious that this was not the final version.

Over all, the change would means:
- More flexible routing
- Flutter-like structure, with smaller component each having a specific role
- The possibility to expand the functionality of this package easily
- Stronger auto-complete

According to mean, this would fix the biggest issue there is: **VStack and VChild now contain 17 attributes, the combination are not always possible**. And this is before adding declarative routing to this package. My idea would be to break down VStack and VChild into several smaller components.

Here is a small example illustrating the idea:

**OLD METHOD**

```dart
  VRouter(
    routes: [
      VStacked(
        path: '/login',
        widget: LoginWidget(),
      ),
      VStacked(
        widgetBuilder: (vChild) => MyScaffold(vChild),
        subroutes: [
          VChild(
            path: '/home',
            widget: HomePage(),
          ),
          VChild(
            path: '/settings',
            widget: SettingsPage(),
          ),
        ]
      ),
    ],
  );
```

**NEW METHOD**
```dart
VRouter(
  route: VSwitcher(
    subroutes: [
      VPath(
        path: '/login',
        subroute: VWidget(
          widget: LoginPage(),
        ),
      ),
      VChildGetter(
        widgetBuilder: (vChild) => MyScaffold(vChild)
        subroute: VSwitcher(
          subroutes: [
            VPath(
              path: '/home',
              subroute: VWidget(
                widget: HomePage(),
              ),
            ),
            VPath(
              path: '/settings',
              subroute: VWidget(
                widget: SettingsPage(),
              ),
            ),
          ]
        ),
      ),
    ],
  ),
);
```
If you count, there is 2 more classes in the new version but each class is very specialized.  This means that you get easier auto completion, and we could easily add new classes to extend the functionalities.

Something interesting is for example to change the path matching algorithm to be more like the react router one, allowing more freedom in the stacking (currently you can't get this [kind of app](https://apps.apple.com/hr/app/github/id1477376905) for example)

Tell me what you think ! Maybe I am completely wrong but I feel like this would be a huge plus for VRouter.

**Comment by orenagiv** [comment]
I think it would be awesome! 👍

**Comment by orenagiv** [comment]
I've noticed that if we redirect directly into a subroute - then there is an issue with the screen transition (the exiting screen remains partially "on top" of the entering screen, and only after the transition ends - it disappears.

It feels like the above concept will might assist to resolve the issue.

**Comment by lulupointu** [comment]
> I've noticed that if we redirect directly into a subroute - then there is an issue with the screen transition (the exiting screen remains partially "on top" of the entering screen, and only after the transition ends - it disappears.
> 
> It feels like the above concept might assist to resolve the issue.

Can you open a new issue and provide an example app ? I cannot reproduce even when I try to nest VStacked inside sudroutes

**Comment by theweiweiway** [comment]
hey @lulupointu 

Some feedback, actually, I much more prefer the older version. Less nesting, less code. I understand your motivation of wanting to make widgets more specialized so they have less arguments, but I feel like it introduces too much code and nesting

For me, I feel like the problem is that it's not completely clear when to use `vChild` vs `vStacked`. I mean, yes, i do know that I should use `vStacked` when i want to wrap a group of `vChild`s with a scaffold or something. But that part isn't super intuitive for me. 

Ideally, (for me at least), there would be a `VRoute` that replaces `VChild` and `VStacked`, and I could choose whether I want to wrap my child routes with a scaffold or not

just my 2 cents.. but I could be wrong.. It's up to what everyone else thinks too

**Comment by lulupointu** [comment]
Thanks for your opinion. I'm looking into allowing nested router right now, which would obliterate any need for VChild. I have a great prototype working, which feels like VRouter but much simpler. I need to keep working to see if everything is working but here is the kind of thing we could have:

```dart
main() {
  runApp(
    VRouterScope(
      child: VRouter(
        routes: [
          VRouteElement(
            path: '/login',
            child: Login(),
          ),
          VRouteElement(
            path: '*',
            child: MyScaffold(),
          ),
        ],
      ),
    ),
  );
}

class MyScaffold extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: BottomNavigationBar(
        currentIndex: VRouterScope.of(context).url.startsWith('/home') ? 0 : 1,
        items: [
          BottomNavigationBarItem(icon: Icon(Icons.home)),
          BottomNavigationBarItem(icon: Icon(Icons.settings)),
        ],
      ),
      body: VRouter(
        routes: [
          VRouteElement(
            path: 'home',
            child: Home(),
          ),
          VRouteElement(
            path: 'settings',
            child: Settings(),
          ),
        ],
      ),
    );
  }
}
```

This does not change my question about large classes vs nesting though, but it might change the answer to which one is the best option

**Comment by orenagiv** [comment]
I like the new approach. And the latest example seems intuitive and simple.

Quick question though:
Why did you use path: '*'  in the example?
I mean, since the MyScaffold has 'home' and 'settings' routes defined without a leading slash and therefore relative to the parent - I suppose the path of the parent should be simply '/', no?

**Comment by theweiweiway** [comment]
Nice, I really like this!

Sorry , yes I think my answer was a little off-point and sidetracking this issue a bit. 

Looking at the second example more closely, I think I can see the separate elements you are trying to get at. Please correct me if I'm wrong: 
`VSwitcher` to define a set of subroutes where only one can be selected
`VChildGetter` if you want to wrap all of it's children with `MyScaffold` or any other widget
`VPath` to define a path
`VWidget` to declare the widget for that path

What would you think about combining `VPath` and `VWidget`? They feel like they are close enough to be one single widget, especially since widget pages are usually associated with a path. It would also reduce the code like so: 

```
VRouter(
  route: VSwitcher(
    subroutes: [
      VPath(
        path: '/login',
        subroute: LoginPage(),
      ),
      VChildGetter(
        widgetBuilder: (vChild) => MyScaffold(vChild)
        subroute: VSwitcher(
          subroutes: [
            VPath(
              path: '/home',
              subroute:  HomePage(),
            ),
            VPath(
              path: '/settings',
              subroute: SettingsPage(),
            ),
          ]
        ),
      ),
    ],
  ),
);
```

**Comment by lulupointu** [comment]
> I like the new approach. And the latest example seems intuitive and simple.
> 
> Quick question though:
> Why did you use path: '*' in the example?
> I mean, since the MyScaffold has 'home' and 'settings' routes defined without a leading slash and therefore relative to the parent - I suppose the path of the parent should be simply '/', no?

Yes this is one of the things that have to change (which I mentioned is my original post here) and which is the way path are matched.
The thing is, MyScaffold has to be displayed with any subpath (apart from '/login'), hence the need for a "*".
Also in a nested VRouter, every path is relative to the parent path. Here the parent path is basically nothing but it's worth noting.

Considering how the router knows how to pick between "*" and "/login", it is thanks to an algorithm of path ranking.

**Comment by lulupointu** [comment]
> Nice, I really like this!
> 
> Sorry , yes I think my answer was a little off-point and sidetracking this issue a bit.
> 
> Looking at the second example more closely, I think I can see the separate elements you are trying to get at. Please correct me if I'm wrong:
> `VSwitcher` to define a set of subroutes where only one can be selected
> `VChildGetter` if you want to wrap all of it's children with `MyScaffold` or any other widget
> `VPath` to define a path
> `VWidget` to declare the widget for that path
> 
> What would you think about combining `VPath` and `VWidget`? They feel like they are close enough to be one single widget, especially since widget pages are usually associated with a path. It would also reduce the code like so:
> 
> ```
> VRouter(
>   route: VSwitcher(
>     subroutes: [
>       VPath(
>         path: '/login',
>         subroute: LoginPage(),
>       ),
>       VChildGetter(
>         widgetBuilder: (vChild) => MyScaffold(vChild)
>         subroute: VSwitcher(
>           subroutes: [
>             VPath(
>               path: '/home',
>               subroute:  HomePage(),
>             ),
>             VPath(
>               path: '/settings',
>               subroute: SettingsPage(),
>             ),
>           ]
>         ),
>       ),
>     ],
>   ),
> );
> ```

Yes I agree, this would even be more true with the new implementation. Currently having a null path is useful (hence the fact that I created VPath), but if I can implement the nesting VRouter, the path algorithm would change so a null path won't be accepted.

Though you can see were this is going. If I eliminate the need for VChildGetter and VPath, the only interesting part remaining would be to know if I still give VWidget beforeLeave, ... or if I abstract this in a VGuard for example

**Comment by theweiweiway** [comment]
I see I see, lll be on the lookout for the new version then - I really like where this is going!

**Comment by daniel-sullivan** [comment]
I think this is a great idea. One thing if I might add is consideration for nesting and provider scope. In our case we have the following:

```
VStacked(
                path: '/product',
                widget: ChangeNotifierProvider(
                  create: (_) => ProductService(),
                  child: ProductFindPage(),
                ),
                subroutes: ProductRoutes().routes,
            )
```

In the routes file:
```
class ProductRoutes {
  final routes = [
    VStacked(
      path: ':id',
      widget: ProductMainPage()
    )
  ];
}
```

It would be ideal being able to use the same ChangeNotifierProvider for an entire branch of the tree. At the moment, it's cut-off since routes and subroutes only support VRouteElement types.

There may be a way to do with is the existing version but haven't been able to locate in the documentation and we're still quite new to Flutter development so aren't really equipped to go digging into the implementation details.

**Comment by lulupointu** [comment]
I don't really understand what you mean by nesting. You mean nesting a VRouter?

Concerning having scoped provider, it is not currently possible and would only be if I allow nested VRouter. This brings some issues but I am concidering it. 

All things considered though, routing should not guide the state but the url should. The idea is to have a unidirectional flow: url->state->UI. Please check [this issue](https://github.com/flutter/uxr/issues/15) where some people are currently discussing these concepts.
if you comply to that.

I'm not sure this answers you question, if it does not please feel free to open a new issue and elaborate!

**Comment by lulupointu** [comment]
Hi 

I took into account the feedback you gave me (thanks!) and after a lot of rewriting I came up with the following implementation:

<details><summary>Click to show sample code</summary>

```dart
VRouter(
  debugShowCheckedModeBanner: false, // VRouter acts as a MaterialApp
  mode: VRouterModes.history, // Remove the '#' from the url
  routes: [
    VWidget(
      path: '/login',
      widget: LoginWidget(),
      subroutes: [
        VGuard(
          beforeEnter: (_) async => print('beforeEnter'),
          subroutes: [
            VNester(
              path: null,
              widgetBuilder: (child) => MyScaffold(child),
              nestedRoutes: [
                VWidget(
                  path: '/settings',
                  widget: InfoWidget(),

                  // Custom transition
                  buildTransition: (animation, ___, child) {
                    return ScaleTransition(
                      scale: animation,
                      child: child,
                    );
                  },
                ),
                VWidget(
                  path: '/profile/:username',
                  // :username is a path parameter and can be any value
                  name: 'profile',
                  // We also give a name for easier navigation
                  widget: ProfileWidget(),

                  // The path '/profile' might also match this path
                  // In this case, we must handle the empty pathParameter
                  aliases: ['/profile'],
                ),
              ],
            ),
          ],
        ),
      ],
    ),
    // This redirect every unknown routes to /login
    VRouteRedirector(
      redirectTo: '/login',
      path: r':_(.*)',
    ),
  ],
)
```

</details>

Here is an overview of the changes:
1. No more `VChild` and `VStack`, but `VWidget` and `VNester`.
This is not only easier to understand but I also more versatile (path are no longer linear since `VNester` has a `nestedRoutes` AND a `subroutes` parameter).
In any case, code completion will be better so it will be easier to start with.
2. As you see, I separated `VGuard`, in the spirit that started this issue. I also created `VPopHandler` which is like `VGuard` but with `onPop` and `onSystemPop`. While this might not seam like but compared to what I proposed when starting this thread (having even a `VPath`) I think this is a good balance between flexibility and verbosity.
3. The new more flexible system allows me to introduce `VPage` and `VNesterPage`, two `VRouteElements` which are the same as `VWidget` and `VNester` but in which you can create a custom page! 
4. No more messing around with `VRouteElementData` `VRouteData` and `VRouterData` ! Everything is now accessible with `VRouter.of`. The history state is now unique but converted to a Map. This should flatten the learning slope drastically. 


Note that I still have to do many tests to ensure not to introduce bugs but I am at a point where I am confident that this is a viable approach and that it will be released. I also have a lot of documentation and migration guide to write.

Also note that `VRouter` can still not be nested. Allowing such a thing would mean sacrificing some stuff that I am not sure are worth the trade (And also I will need to get some sleep after this release ^^).

I hope you are as exited as I am! Don't hesitate to leave your feedback!

**Comment by biteater** [comment]
Hi,

I'm just starting with flutter and while trying to understand the Navigation 2 thingy, I found VRouter in the corresponding discussion. After having a little experience with routing in Vue, I really like this idea. Regarding my comments, bear in mind, that I have close to zero experience with flutter, so don't hesitate to tell me if I'm getting it wrong :)

`VGuard`: That looks like a really useful feature and I certainly know where I will need it. And the name is great. Go for it!

`VNester`: Maybe I might suggest a different name. How about `VWrap` or `VWrapper`, as it is wrapping around the nested routes. Additionally it might be useful to rename the the parameter _nestedRoutes_ into _subroutes_. It's just what they are, and you wouldn't have to remember that the "same kind of stuff" goes into a different parameter here.

`VWidget`: As far as I have seen so far, Widgets in flutter do contain the widgets in the parameter "_child_", here it is a different one. If this is "a widget" it might be easier to use if it uses the same name for "the same thing". Maybe I don't yet understand exactly what widgets are, but what I got so far is that widgets are elements that are used for the UI, so they are (more or less) visible elements or building blocks for the user interface. 

I personally would not consider the definition of the navigation itself as a visible element. To clarify the purpose, a different name might help. I liked the name `VPath` for path-elements, but it is not exactly what this is, is it? To me it seems that `VChild` or `VRoute` might come a little closer to the purpose of this element, I would prefer the latter.

Anyhow, this is just cosmetics and does not really change the usage, but might be helpful to understand and use it even more intuitively :)

Continue this project and create something great!

Hagen

**Comment by lulupointu** [comment]
Thanks for taking the time to write this feedback!

I will answer each of you point separately:

- `VGuard`: thanks, I will definitely keep this name.

- `VNester`: I agree that this might not be a beautiful name but in Flutter we always talk about *nesting widgets* so I think this is appropriate. **However** I do understand that using `subroute` would feel more consistent but here is the thing: `subroute` is a valid parameter ! Imagine you want to achieve the following structure

```
|---------------------------------------------|
|                Settings                     | 
|---------------------------------------------|
|                                             |
|             Scaffold                        |
|                 |_ |------------------|     |
|                    |  ScaffoldChild   |     |
|                    |------------------|     |
|                                             |
|---------------------------------------------|
|                Home                         | 
|---------------------------------------------|
```

This is now possible using the following code:

```dart
VRouter(
  routes: [
    VWidget(
      path: '/home',
      widget: HomeWidget(),
      stackedRoutes: [
        VNester(
          path: '/app',
          widgetBuilder: (child) => MyScaffold(child),
          nestedRoutes: [
            VWidget(
              path: ':_(.*)',
              widget: ScaffoldChildWidget(),
            ),
          ],
          stackedRoutes: [
            VWidget(
              path: 'settings',
              widget: SettingsWidget(),
            ),
          ],
        ),
      ],
    ),
  ],
)
```

You might notice things:
1. I just changed `subroutes` to `stackedRoutes`, this is because a lot of people familiar with flutter did not understand that the route which is stacked is not the following one in the list of subroute but really the one which is *in* `subroutes`. Using `stackedRoutes` makes it very clear imo but tell me what you think.
2. Using `nestedRoutes` is necessary because, as per the example above, a `VNester` **must** have `nestedRoutes` AND `stackedRoutes`. So maybe the naming can be better but in any case it has to be different.

The point number 2 brings me back to you remark on the name `VNester`: Since I must have an other name, I think it is very intuitive to have `VNester` and `nestedRoute` (matching name). However I would name this `VWrapper` and `wrappedRoutes` if I followed your logic. I would say that the term `nest` is still more aligned to the flutter philosophy than `wrap` but I would love to hear what others think about it.

- `VWidget`: The source of confusion I think is that `VRouteElement`s are **not** widgets. I know, that's a bummer... Everything is supposed to be a widget in Flutter but there are not, and they can't be because their life cycle is very different from the one of widgets.
This causes some confusion but this confusion makes it easy for new comers to understand how they work: just like widgets, nest them and there you go! 
This explains the choice of `widget` instead of `child`. This is a tough choice and I'm not 100% it is the best, I would love to know what the others think about it!

- `VPath`: I decide not to use this element and to include the path in `VWidget` or `VNester` directly (as @theweiweiway suggested). This might lead to more verbosity *sometimes* but will lead to less nesting (i.e. clearer) *most of the time*.

Thanks again for you comment! It's great to have feedback from new flutter devs to!

**Comment by lulupointu** [comment]
Hi,

v1.1 is coming soon, I just need to update the vrouter.dev website (and certainly other tiny things I forgot about).

Anyhow, you can access the new version with the [dev channel](https://github.com/lulupointu/vrouter/tree/dev), you can easily get a grasp of the new version by looking at the [README](https://github.com/lulupointu/vrouter/blob/dev/README.md). I added a migration section at the bottom of the README so I am very interested in your opinion here.

**Comment by orenagiv** [comment]
Awesome news @lulupointu!
Thanks so much! 🙏

**Comment by lulupointu** [comment]
:fire:  v1.1 is out! :fire:

See the migration guide if you need, it is at the end of the README.
Also [vrouter.dev](https://vrouter.dev/) has been updating so check it out :+1: 

I can't wait to have your feedback, don't hesitate to open a new issue if needed!

**Comment by orenagiv** [comment]
Have just migrated :)
Awesome work @lulupointu !
THANK YOU.

**Comment by lulupointu** [comment]
@orenagiv Could you try the following snippet on Android or iOS **with `vrouter: 1.1.0+4`** in pubspec.yaml and tell me if the app is put in the background when you press the button?

```dart
import 'package:flutter/material.dart';
import 'package:vrouter/vrouter.dart';

main() {
  runApp(
    VRouter(
      routes: [
        VWidget(
          path: '/',
          widget: Builder(
            builder: (context) => TextButton(
              child: Text('Click to pop'),
              onPressed: () => Navigator.of(context).pop(),
            ),
          ),
        ),
      ],
    ),
  );
}
```

**Comment by orenagiv** [comment]
Hey @lulupointu 
Tested on iOS (v14.4) with vrouter 1.1.0+4 and yes - clicking the "click to pop" button moves the app to the background.
I'll test on Android a bit later today.

**Comment by lulupointu** [comment]
Great! If this works on ios I don't think there should be any issue on Android so don't bother.

Thank you very much for testing!

**Comment by orenagiv** [comment]
NP :)
Checked on Android v9 just to be sure - works too.
